app/src/main/java/ch/epfl/sdp/appart/database/local/LocalDatabase.java

```diff
@@ -0,0 +1,1055 @@
+     * given to this function. We opted to do this, this way because we do
+     * not want the local database to be aware if Firebase.
+     *
+     * @param loadProfilePic the function that performs the write on disk of
+     *                       the profile picture of the user
+     * @param futureSuccess  a completable future that indicates if the
+     *                       operation succeeded or not.
+     */
+    private void writeProfilePic(Function<String,
+            CompletableFuture<Void>> loadProfilePic,
+                                 CompletableFuture<Void> futureSuccess) {
+
+        CompletableFuture<Void> futureProfilePic =
+                loadProfilePic.apply(LocalDatabasePaths.userProfilePic(getCurrentUser().getUserId()));
+        futureProfilePic.thenAccept(arg -> futureSuccess.complete(null));
+        futureProfilePic.exceptionally(e -> {
+            futureSuccess.completeExceptionally(e);
+            return null;
+        });
+
+        futureSuccess.complete(null);
+    }
+
+    /**
+     * This function performs the writing of a complete ad into local storage
+     * . It will create a folder for the added user only if it doesn't
+     * already exists (in which case it is updated). It will do the same for
+     * the ad. It also performs the writing of all the images. The photos of
+     * ad, the panoramas of the ad and the profile picture of the user (if he
+     * doesn't uses the default one).
+     *
+     * @param adId           the id of the ad
+     * @param cardId         the id of the card
+     * @param ad             the ad
+     * @param user           the user who posted the ad
+     * @param adPhotos       list of bitmaps representing the photos of the ad
+     * @param panoramas      list of panoramas representing the panoramas of
+     *                       the ad
+     * @param loadProfilePic the function that will perform the writing on
+     *                       disk of the user's profile picture. It receives
+     *                       the path representing the location of the
+     *                       profile picture and returns a completable future
+     *                       to indicate if the operation succeeded or not.
+     * @return a completable future that indicates if the operation succeeded
+     * or not
+     */
+    public CompletableFuture<Void> writeCompleteAd(String adId,
+                                                   String cardId, Ad ad,
+                                                   User user,
+                                                   List<Bitmap> adPhotos,
+                                                   List<Bitmap> panoramas,
+                                                   Function<String,
+                                                           CompletableFuture<Void>> loadProfilePic) {
+
+        if (adId == null || cardId == null || ad == null || user == null || adPhotos == null || panoramas == null)
```

Comment:
We should do static methods like `checkAd(Ad ad)` and this way we will have more fine-grained control over which argument can or cannot be null or cannot take a value. We could even do checks with regex over the String fields. I let it here.

Reply:
For the comment before this one, it doesn't require the context as it just calls the function loadProfilePic but doesn't know anything about it.